### PR TITLE
[backport] Fix fp16 issue in batch normalization + Keep backward compatibility on cupy.cudnn.batch_normalization_forward_training

### PR DESCRIPTION
**Pull request by toslunar**
Backport of #6323 and #6497.

Merge cupy/cupy#2094 first.

### CODE REVIEW

**Comment by hvy** [comment]
@takagi Can you take a look at this backport?

**Comment by takagi** [comment]
Now cupy/cupy#2094 is merged.

**Comment by takagi** [comment]
LGTM!
